executor/executor.go

```diff
@@ -2318,3 +2331,470 @@ func isWeakConsistencyRead(ctx sessionctx.Context, node ast.Node) bool {
+	done       bool
+	is         infoschema.InfoSchema
+	err        error
+	hasError   *atomic.Bool
+	wg         sync.WaitGroup
+	contextCtx context.Context
+}
+
+// Open implements the Executor Open interface.
+func (e *FastCheckTableExec) Open(ctx context.Context) error {
+	if err := e.baseExecutor.Open(ctx); err != nil {
+		return err
+	}
+
+	e.done = false
+	e.contextCtx = ctx
+	return nil
+}
+
+type checkIndexTask struct {
+	indexOffset int
+}
+
+type checkIndexWorker struct {
+	sctx       sessionctx.Context
+	dbName     string
+	table      table.Table
+	indexInfos []*model.IndexInfo
+	e          *FastCheckTableExec
+	ctx        context.Context
+}
+
+type groupByChecksum struct {
+	bucket   uint64
+	checksum uint64
+	count    int64
+}
+
+func getCheckSum(ctx context.Context, se sessionctx.Context, sql string) ([]groupByChecksum, error) {
+	ctx = kv.WithInternalSourceType(ctx, kv.InternalTxnAdmin)
+	rs, err := se.(sqlexec.SQLExecutor).ExecuteInternal(ctx, sql)
+	if err != nil {
+		return nil, err
+	}
+	defer func(rs sqlexec.RecordSet) {
+		err := rs.Close()
+		if err != nil {
+			logutil.BgLogger().Error("close record set failed", zap.Error(err))
+		}
+	}(rs)
+	rows, err := sqlexec.DrainRecordSet(ctx, rs, 256)
+	if err != nil {
+		return nil, err
+	}
+	checksums := make([]groupByChecksum, 0, len(rows))
+	for _, row := range rows {
+		checksums = append(checksums, groupByChecksum{bucket: row.GetUint64(1), checksum: row.GetUint64(0), count: row.GetInt64(2)})
+	}
+	return checksums, nil
+}
+
+// HandleTask implements the Worker interface.
+func (w *checkIndexWorker) HandleTask(task checkIndexTask) {
+	defer w.e.wg.Done()
+	idxInfo := w.indexInfos[task.indexOffset]
+	bucketSize := int(CheckTableFastBucketSize.Load())
+
+	ctx := kv.WithInternalSourceType(w.e.contextCtx, kv.InternalTxnAdmin)
+
+	trySaveErr := func(err error) {
+		if setNewErr := w.e.hasError.CompareAndSwap(false, true); setNewErr {
+			w.e.err = err
+		}
+	}
+
+	se, err := w.e.base().getSysSession()
+	se.GetSessionVars().OptimizerUseInvisibleIndexes = true
+	if err != nil {
+		trySaveErr(err)
+		return
+	}
+	defer func() {
+		se.GetSessionVars().OptimizerUseInvisibleIndexes = false
+		w.e.base().releaseSysSession(ctx, se)
+	}()
+
+	var pkCols []string
+	var pkTypes []*types.FieldType
+	var primaryIdx *model.IndexInfo
+	switch {
+	case w.e.table.Meta().IsCommonHandle:
+		pkColsInfo := w.e.table.Meta().GetPrimaryKey().Columns
+		for _, colInfo := range pkColsInfo {
+			colStr := colInfo.Name.O
+			pkCols = append(pkCols, colStr)
+		}
+		primaryIdx = tables.FindPrimaryIndex(w.table.Meta())
+	case w.e.table.Meta().PKIsHandle:
+		pkCols = append(pkCols, w.e.table.Meta().GetPkName().O)
+		primaryIdx = tables.FindPrimaryIndex(w.table.Meta())
+	default: // support decoding _tidb_rowid.
+		pkCols = append(pkCols, model.ExtraHandleName.O)
+	}
+
+	// CheckSum of (handle + index columns).
+	var md5HandleAndIndexCol strings.Builder
+	md5HandleAndIndexCol.WriteString("crc32(md5(concat(")
```

Comment:
I don't think `concat()` is appropriate because `concat('ab', 'c')` is identical to `concat('a', 'bc')`. How about adding column IDs to reduce the force-positive cases?